src/ndb.c

```diff
@@ -0,0 +1,369 @@
+      if (data_write_ok && md_write_ok)
+        log_info("Data and metadata were written to the NDB file");
+      else {
+        if (!data_write_ok)
+          log_error("Error writing data to the NDB file");
+        if (!md_write_ok)
+          log_error("Error writing metadata to the NDB file");
+      }
+    }
+  }
+}
+
+enum ndb_op_code do_nv_writes(bool *data_write_ok, bool *md_write_ok)
+{
+  ndb_element_metadata_t* md = NULL;
+  enum ndb_op_code ret;
+  int changed;
+
+  if (NDB_ERR_TIMEOUT == ndb_wq_get(&md)) {
+    return NDB_ERR_TIMEOUT;
+  } else {
+    if (!do_writes_to_flash) {
+      *data_write_ok = *md_write_ok = true;
+      return NDB_ERR_NONE;
+    }
+    ndb_element_metadata_t md_copy;
+    ndb_element_t buf;
+    ndb_lock();
+    memcpy(&buf, md->data, md->data_size);
+    memcpy(&md_copy, md, sizeof(ndb_element_metadata_t));
+    ndb_unlock();
+    ret = ndb_write_file_data(md->file, md->data_size * md->index, &buf,
+                              md->data_size);
+    *data_write_ok = NDB_ERR_NONE == ret;
+    if (data_write_ok) {
+      ndb_lock();
+      changed = memcmp(md->data, &buf, md->data_size);
+      ndb_unlock();
+      if (!changed) {
+        /*
+         * Data hasn't changed while i/o operation was running -
+         * mark as saved and write metadata to NVM.
+         * */
+        md->nv_data.state &= ~NDB_IE_DIRTY;
+        cfs_offset_t offset = md->data_size * md->n_elements
+            + sizeof(ndb_element_metadata_nv_t) * md->index;
+        ret = ndb_write_file_data(md->file, offset, &md_copy,
+                                  sizeof(ndb_element_metadata_nv_t));
+        *md_write_ok = NDB_ERR_NONE == ret;
+      } else {
+        /* Data has changed - schedule saving once again */
+        md->nv_data.state |= NDB_IE_DIRTY;
+        ndb_wq_put(md);
+      }
+    }
+  }
+
+  return NDB_ERR_NONE;
+}
+
+void do_sbp_updates()
+{
+  ndb_p1_sbp_update();
+  ndb_p3_sbp_update();
+}
+
+/**
+ * Opens NDB file that stores information elements of certain type.
+ * This function checks if version of the file matches the passed one
+ * and if it doesn't creates empty file automatically.
+ */
+enum ndb_op_code ndb_open_file(ndb_file_t *file, char *version)
+{
+  char ver[MAX_NDB_FILE_VERSION_LEN];
+
+  if (!do_writes_to_flash) {
+    return NDB_ERR_NONE;
+  }
+
+  file->fh = cfs_open(file->name, CFS_READ + CFS_WRITE);
+  if (file->fh < 0) {
+    cfs_coffee_reserve(file->name, file->expected_size);
+    cfs_coffee_configure_log(file->name, 256 * 2, 256);
+    file->fh = cfs_open(file->name, CFS_READ + CFS_WRITE);
+    if (file->fh < 0) {
+      return NDB_ERR_FILE_IO;
+    }
+  }
+
+  if (cfs_read(file->fh, ver, sizeof(ndb_file_version))
+      == sizeof(ndb_file_version)) {
+    if (memcmp(&ver, version, sizeof(ndb_file_version)) == 0) {
+      return NDB_ERR_NONE;
+    } else {
+      cfs_close(file->fh);
+      cfs_remove(file->name);
+      return ndb_open_file(file, version);
+    }
+  }
```

Comment:
If file contains less data than sizeof(ndb_file_version) we just reuse that file as it was opened successfully for read/write access. We write there version signature and it's equivalent to creating a new file and doing the same write. Am I wrong somewhere?